src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes/ApiControllerCodeFixer.cs

```diff
@@ -0,0 +1,100 @@
+        {
+            // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            if (root is null)
+            {
+                return;
+            }
+
+            var node = root.FindNode(context.Span, false, true);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            // Register a code action that will invoke the fix.
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    CodeFixResources.ApiControllerTitle,
+                    cancellationToken => ReplaceBaseClass(context.Document, node, cancellationToken),
+                    nameof(CodeFixResources.ApiControllerTitle)),
+                context.Diagnostics);
+        }
+
+        /// <summary>
+        /// Built with the assumption that the <paramref name="node"/> is a IdentifierNameSyntax or QualifiedNameSyntax that indicates an ApiController BaseType.
+        /// </summary>
+        /// <param name="document">The document that contains a class to fix.</param>
+        /// <param name="node">A class that is ApiController.</param>
+        /// <param name="cancellationToken">May request cancellation.</param>
+        /// <returns>A document where at least one ApiController was removed.</returns>
+        private async Task<Document> ReplaceBaseClass(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var mvcBaseType = generator.QualifiedName(
+                CreateMvcNamespace(generator),
+                generator.IdentifierName(GoodClassName))
+                .WithAdditionalAnnotations(Simplifier.Annotation, Simplifier.AddImportsAnnotation);
+
+            var baseType = GetBaseType(node, generator);
+
+            var newRoot = generator.ReplaceNode(root, baseType, mvcBaseType);
+
+            var newDocument = document.WithSyntaxRoot(newRoot);
+
+            newDocument = await ImportAdder.AddImportsAsync(newDocument, Simplifier.AddImportsAnnotation, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return newDocument;
+        }
+
+        private static SyntaxNode GetBaseType(SyntaxNode node, SyntaxGenerator generator)
+        {
+            var types = generator.GetBaseAndInterfaceTypes(node);
+            if (types == null || types.Count == 0)
+            {
+                types = generator.GetBaseAndInterfaceTypes(node.Parent);
+            }
+
+            return types[0];
```

Comment:
note: This may not work in the cae of a partial type.  Consider a partial type like so:

```
partial class Foo {
}

partial calss Foo : TheBaseType {
}
```

If you are running on the part that has no base types, this will index out of bounds.  A way to address this is to make sure on the analyzer side that this helper returns a non-null node.  as per our convo, you could also then just store that node in the AdditionalLocations part of hte diagnostic to make it easy to find in the fixer.

Reply:
updated unit test to include this scenario and reworked the analyzer to detect it